Framework/Geometry/inc/MantidGeometry/Instrument/StructuredDetector.h

```diff
@@ -0,0 +1,197 @@
+namespace Geometry {
+/**
+*  StructuredDetector is a type of CompAssembly, an assembly of components.
+*  It is designed to be a more efficient way of defining instruments which
+*  are topologically regular but geometrically irregular.
+*
+* @class StructuredDetector
+* @brief Assembly of Detector objects with topological regularity and
+geometrical irregularity
+* @author Lamar Moore, ISIS
+* @date 07-March-2016
+
+Copyright &copy; 2016-17 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
+National Laboratory & European Spallation Source
+
+This file is part of Mantid.
+
+Mantid is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation; either version 3 of the License, or
+(at your option) any later version.
+
+Mantid is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+File change history is stored at: <https://github.com/mantidproject/mantid>
+Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+class MANTID_GEOMETRY_DLL StructuredDetector : public CompAssembly,
+                                               public IObjComponent {
+public:
+  /// String description of the type of component
+  std::string type() const override { return "StructuredDetector"; }
+
+  /// Matches name to Structured Detector
+  static bool compareName(const std::string &proposedMatch);
+
+  //! Empty constructor
+  StructuredDetector();
+
+  //! Constructor with a name and parent reference
+  StructuredDetector(const std::string &name, IComponent *reference = nullptr);
+
+  //! Parametrized constructor
+  StructuredDetector(const StructuredDetector *base, const ParameterMap *map);
+
+  /// Create all the detector pixels of this rectangular detector.
+  void initialize(int xpixels, int ypixels, const std::vector<double> &x,
+                  const std::vector<double> &y, int idstart,
+                  bool idfillbyfirst_y, int idstepbyrow, int idstep = 1);
+
+  //! Make a clone of the present component
+  IComponent *clone() const override;
+
+  boost::shared_ptr<Detector> getAtXY(const int X, const int Y) const;
+
+  detid_t getDetectorIDAtXY(const int X, const int Y) const;
+  std::pair<int, int> getXYForDetectorID(const int detectorID) const;
+
+  std::vector<double> const &getXValues() const;
+  std::vector<double> const &getYValues() const;
+
+  void setColors(const std::vector<int> &r, const std::vector<int> &g,
+                 const std::vector<int> &b) const;
+
+  std::vector<int> const &getR() const;
+  std::vector<int> const &getG() const;
+  std::vector<int> const &getB() const;
+
+  int xpixels() const;
```

Comment:
It's not clear form the name that it's the number of pixels in the x-direction. Should the type be `size_t`?